### PR TITLE
baetyl-779: Add CAS when update desire

### PR DESCRIPTION
**Pull request by hannatao**

### CODE REVIEW

File: service/node.go [hunk 1]
```diff
@@ -273,17 +282,33 @@ func (n *nodeService) UpdateReport(namespace, name string, report specV1.Report)
...
+		if newShadow == nil {
+			f(shadow, app)
+			return n.createShadow(shadow.Namespace, shadow.Name, shadow.Desire, nil)
+		}
 
-	return n.updateDesire(shadow, desire)
+		// Refresh desire in shadow by app
+		f(newShadow, app)
+		updatedShadow, err := n.shadow.UpdateDesire(newShadow)
+		if err == nil || err.Error() != "invalid version" {
```
Comment (allen-lj) [inline service/node.go:303]: add return update result（success or failed）and check the update result in the if statement

File: service/node.go [hunk 2]
```diff
@@ -273,17 +282,33 @@ func (n *nodeService) UpdateReport(namespace, name string, report specV1.Report)
 }
 
 // UpdateDesire Update Desire
-func (n *nodeService) UpdateDesire(namespace, name string, desire specV1.Desire) (*models.Shadow, error) {
-	shadow, err := n.shadow.Get(namespace, name)
-	if err != nil {
-		return nil, err
-	}
+// Parameter f can be RefreshNodeDesireByApp or DeleteNodeDesireByApp
+func (n *nodeService) UpdateDesire(shadow *models.Shadow, app *specV1.Application, f func(*models.Shadow, *specV1.Application)) (*models.Shadow, error) {
```
Comment (allen-lj) [inline service/node.go:286]: the shadow parameter is unnecessary

File: service/node.go [hunk 3]
```diff
@@ -141,6 +141,15 @@ func (n *nodeService) Update(namespace string, node *specV1.Node) (*specV1.Node,
 	return res, nil
 }
 
+//List get shadow list
+func (n *nodeService) ShadowList(namespace string, listOptions *models.ListOptions) (*models.ShadowList, error) {
+	list, err := n.node.ListNode(namespace, listOptions)
+	if err != nil {
+		return nil, err
+	}
+	return n.shadow.List(namespace, list)
```
Comment (allen-lj) [inline service/node.go:150]: showList maybe unnecessary

File: service/node.go [hunk 4]
```diff
@@ -480,10 +482,24 @@ func refreshNodeDesireByApp(node *specV1.Node, app *specV1.Application) {
 		apps[idx].Version = app.Version
 	}
 
-	node.Desire.SetAppInfos(app.System, apps)
+	shadow.Desire.SetAppInfos(app.System, apps)
 
 }
 
+func DeleteNodeDesireByApp(shadow *models.Shadow, app *specV1.Application)  {
+	if shadow.Desire != nil {
```
Comment (chensheng0) [inline service/node.go:490]: if xx == nil {
return
}

**Review by chensheng0 — APPROVED** [review]

**Pull request merged**